src/glyph/mod.rs

```diff
@@ -104,6 +104,28 @@ impl Glyph {
         }
     }
 
+    /// Returns boolean value indicating whether [`Glyph`] is defined with one or
+    /// more [`Component`].
+    pub fn has_component(&self) -> bool {
+        !self.components.is_empty()
+    }
+
+    /// Returns a usize that represents the number of [`Component`] defined on the Glyph.
+    pub fn component_number(&self) -> usize {
+        self.components.len()
+    }
+
+    /// Returns boolean indicating the presence of one or more [`Component`] with base
+    /// glyph name `basename`.
+    pub fn has_component_with_base(&self, basename: &str) -> bool {
+        self.components.iter().any(|x| *x.base == *basename)
+    }
+
+    /// Return a Vec of immutable [`Component`] references filtered by base glyph name.
+    pub fn get_components_with_base(&self, basename: &str) -> Vec<&Component> {
+        self.components.iter().filter(|x| *x.base == *basename).collect::<Vec<&Component>>()
+    }
```

Comment:
Instead of returning a `Vec` (and allocating) I would prefer to return an `Iterator` and then give the caller the choice of allocating if necessary.

This ends up being slightly complicated for subtle lifetime reasons: we're returning an iterator that itself returns references to items owned by self, but the iterator *itself* also needs to outlive `basename`, since we need to compare with `basename` each time we `filter`.

The final complication is that we need to express that the two lifetimes are not independent: items borrowed from `self` must be valid for at least as long as the reference to `basename`. This means we end up with something like the following (it's possible there is a more concise way of expressing this, but I couldn't think of one):

```suggestion
    pub fn get_components_with_base<'b, 'a: 'b>(&'a self, basename: &'b str) -> impl Iterator<Item=&'a Component> + 'b {
        self.components.iter().filter(move |x| *x.base == *basename)
    }
```

Reply:
I started on this path but chickened out when I needed to add the `move` in order to make it work.  Does that move ownership out of `self.components` and can `self.components` be reused for access to the components data after  `get_components_with_base` execution?

Reply:
I suspect someone else could explain this better, but basically the move only implicates the items referenced *inside* the closure that are defined outside the closure's scope, so in this case the `basename` argument. It's definitely confusing in this case, but I would take it to mean that the whole `&str` is moved into the closure, instead of being referenced (as an `&&str`?)